feat: add `slice:custom-screenshot:upsert` and `prismic:push` hook

lihbr opened the pull request:
## `slice:custom-screenshot:upsert`

The goal of this hook is to tell plugins the custom screenshot is available at *location* (`customScreenshotAbsolutePath`). Adapters are then meant to save it using their preferred method.

Comments:
1. Is the hook name alright? `create` is not really accurate in representing what it does, same for `update`. Maybe there's a better name though~
2. The hook receives the location under the key `customScreenshotAbsolutePath`:
   1. Do we want it absolute? Or relative (from root) would be better? I think absolute makes more sense as it could be written in any `tmp` directory on the OS
   2. Since it's a filesystem path I feel like it's important to highlight the relativeness or the absoluteness of the path in the property name to not create confusion/inconsistencies issues.
3. The `slice:read` hook has been updated to return an optional `customScreenshotAbsolutePath` property should a custom screenshot be available:
   1. Same comments as #2
   2. Since it's described as a path, what if the plugin were to save it elsewhere than in the filesystem (namely, online)? I don't think we want Slice Machine to download URLs and maybe plugins using this pattern should download it in a tmp directory if so.

## `prismic:push`

The goal of this hook is to represent interactions between Slice Machine and the custom-type API.
It can receive slices or custom types identifiers alongside the action performed (create, update, delete, decided after custom-type API request types: https://prismic.io/docs/technologies/custom-types-api#add-custom-types-post-requests)

lihbr commented:
Not so related to this PR, but is there a reason we used capitalized string constants there?
- https://github.com/lihbr/prismic-slice-machine-plugin-exploration/pull/6/files#diff-805597452eb9ccde160b18873fa87ec57f00b9503d602361db3b389b1afb565eL230-L233
- https://github.com/lihbr/prismic-slice-machine-plugin-exploration/blob/master/packages/kit/src/types.ts#L288-L291

I feel like lowercase is more standard 🤔

angeloashmore commented:
## `slice:custom-screenshot:upsert`

> 1. Is the hook name alright? create is not really accurate in representing what it does, same for update. Maybe there's a better name though~

I think we could remove the `custom-` prefix. I'm assuming this would handle all screenshots, even ones taken with Puppeteer.

"upsert" is more technically accurate, but it introduces another type of action. We have "create," "read," "update," and "delete" (CRUD) for Slices and Custom Types. How can we keep new actions aligned as closely as possible?

1. Do we remove "create" and "update" from Slice/Custom Types and replace it with "upsert"?
2. Or do we replace "screenshot:upsert" with "screenshot:create" and "screenshot:update"?

I'm in favor of Option 1. It means we can better handle Slice/Custom Type ID and name changes (should that become a feature). It means a little more complexity in the `upsert` hooks, but with the benefit of handling more edge cases.

> 2. The hook receives the location under the key `customScreenshotAbsolutePath`:
>    1. Do we want it absolute? Or relative (from root) would be better? I think absolute makes more sense as it could be written in any tmp directory on the OS
>    2. Since it's a filesystem path I feel like it's important to highlight the relativeness or the absoluteness of the path in the property name to not create confusion/inconsistencies issues.

Absolute makes sense to me. 👍 Ideally it's always pointed to a tmp directory.

I don't think it's 100% necessary to highlight that it's an absolute path in the property name. It would be important to highlight it in the property's TSDoc, however. I'm okay either way - up to you! I see why it would be helpful.

> 3. The slice:read hook has been updated to return an optional `customScreenshotAbsolutePath` property should a custom screenshot be available:
>    1. Same comments as https://github.com/lihbr/prismic-slice-machine-plugin-exploration/pull/2
>    2. Since it's described as a path, what if the plugin were to save it elsewhere than in the filesystem (namely, online)? I don't think we want Slice Machine to download URLs and maybe plugins using this pattern should download it in a tmp directory if so.

Requiring plugins to return a local file path makes sense to me. Slice Machine will need to be able to load the display these images. I'm not sure how it will do that if the image lives in, for example, `/tmp`. Will it need to copy it into a temporary `./public` directory?

The concept of passing file paths back and forward is predicated on the idea that Slice Machine will upload images to S3 on `prismic:push` (or a similar action). Image upload is effectively the last step in the create/update model flow.

Reminder: screenshot URLs are part of the Slice JSON model. After uploading the S3, the model will need to be updated through the adapter.

It might be helpful if we create a simple flow chart illustrating the flow from Slice create (or update) to pushing to Prismic, including any side-effects.

## `prismic:push`

> The goal of this hook is to represent interactions between Slice Machine and the custom-type API.
It can receive slices or custom types identifiers alongside the action performed (create, update, delete, decided after custom-type API request types: https://prismic.io/docs/technologies/custom-types-api#add-custom-types-post-requests)

This looks good to me. 👍 We'll probably want to check with the Slice Machine team to verify this matches their Custom Type API changes (if a new endpoint is being created).

lihbr commented:
I updated the PR following what we agreed upon, remaining for discussion:

## `slice:custom-screenshot:upsert`

> I think we could remove the custom- prefix. I'm assuming this would handle all screenshots, even ones taken with Puppeteer.

Would it handle all screenshots? Because Slice Machine's taking a screenshot automatically with Puppeteer, so how would we differentiate between the two/not overwrite custom screenshots? From my understanding Puppeteer screenshot would still be inside `.slicemachine`

> "upsert" is more technically accurate, but it introduces another type of action. We have "create," "read," "update," and "delete" (CRUD) for Slices and Custom Types. How can we keep new actions aligned as closely as possible?

I'm not in favor of option 1 as it makes hooks less precise (a 3rd party plugin couldn't hook to `slice:upsert` and differentiate between a creation or update)

I think I prefer option 2 and just roll with "update" considering the custom screenshot is updated from "none" to "something".

> I don't think it's 100% necessary to highlight that it's an absolute path in the property name. It would be important to highlight it in the property's TSDoc, however. I'm okay either way - up to you! I see why it would be helpful.

TSDoc to the rescue 💯, updated.

> Slice Machine will need to be able to load the display these images. I'm not sure how it will do that if the image lives in, for example, `/tmp`. Will it need to copy it into a temporary `./public` directory?

Couldn't Slice Machine API act as a proxy? e.g. `GET /api/:libraryID/:sliceID/:variationID/screenshot` returns either the custom screenshot image, or, if none, the automatic one.

## `prismic:push`

> This looks good to me. 👍 We'll probably want to check with the Slice Machine team to verify this matches their Custom Type API changes (if a new endpoint is being created).

Sure, I'll try to investigate, let's keep it like this for now as experimental, but at least it's there.

angeloashmore commented:
# Screenshots

Here's the flow that I'm thinking of. The following flow happens within Slice Machine and its UI.

<img width="2480" alt="Untitled" src="https://user-images.githubusercontent.com/8601064/176802189-2fdff4bc-559b-4df0-82ac-117fde8da0d0.png">

`syncSliceScreenshot()` updates the `imageUrl` property with a screenshot URL if a new screenshot is detected.

This flow decouples the existing screenshot flow from Slice Machine's Slice model update path. A Slice's screenshot file can be saved to the file system via a "Take Screenshot" button in Slice Machine with Puppeteer (assuming we have a `slice:screenshot:update` action to save it), or the screenshot could be a manually added file.

This flow also decouples screenshots from the file system. Adapters MUST implement `slice:screenshot:read`, `slice:screenshot:update`, and `slice:screenshot:delete` to manage screenshots. The Next.js and Nuxt adapters will manage screenshots by saving to the file system.

We may want to add a `slice:update:before` hook that runs before `slice:update`. It allows actions to be performed before the Slice model is saved. For plugins that manage screenshots, like a Figma plugin, that hook could be how/when it updates the screenshot. Note that this means we'll need an action like `updateScreenshot()` that plugins can call to trigger `slice:screenshot:update`.

---

> I'm not in favor of option 1 as it makes hooks less precise (a 3rd party plugin couldn't hook to slice:upsert and differentiate between a creation or update)

Plugins could know if its updating or creating something by checking the presence of a directory (or DB row, etc.). I think it would be simple. `create` and `update` are basically the same thing (hence the existence of `upsert`). Do you see a benefit to keeping them separated?

---

> Couldn't Slice Machine API act as a proxy? e.g. GET /api/:libraryID/:sliceID/:variationID/screenshot returns either the custom screenshot image, or, if none, the automatic one.

Yup, you're right; the API should be able to proxy an image. Good idea!

lihbr commented:
> Plugins could know if its updating or creating something by checking the presence of a directory (or DB row, etc.). I think it would be simple. create and update are basically the same thing (hence the existence of upsert). Do you see a benefit to keeping them separated?

The adapter could, or plugins that want to manage their own set of "slice things". If you were to create a tracking plugin for example though you wouldn't be able to differentiate between the two because such plugin wouldn't know where's the slice directory.
